### ISSUE TITLE
Enum type inference could use improvement

### ISSUE THREAD

**Issue by bprg** [icsharpcode/ILSpy#1142]
C#
`var value = 0xfffffe0b;`

IL
`IL_0000: ldc.i4 -501`

**Comment by siegfriedpammer** [comment]
Why is this invalid? What would you expect? `-501` and `0xfffffe0b` have the same value, if encoded in binary and `ilasm` accepts both `ldc.i4 -501` and `ldc.i4 0xfffffe0b`. It seems `ildasm` always uses hexadecimal notation for values < 0, but that is just a question of formatting the output value.

`ldc.i4 -504` is not invalid.

**Comment by bprg** [comment]
You're right, But if added hexadecimal value in comment like dotPeek decompiler is great.

`IL_0000: ldc.i4 -501 //0xfffffe0b`

**Comment by bprg** [comment]
Another issue with enum.

Code
```
MyEnum value = MyEnum.Val2;
if (value - MyEnum.Val1 <= 4)
{
   Console.WriteLine(value - MyEnum.Val1);
}

enum MyEnum : uint
{
   Val1 = 0xfffffe0b,
   Val2 = 0xfffffe0f
}
```

Decompile result
```
MyEnum myEnum = MyEnum.Val2;
if ((uint)((int)myEnum - -501) <= 4u)
{
   Console.WriteLine((uint)((int)myEnum - -501));
}
```
